pkg/blockchain/rpcLogStreamer.go

```diff
@@ -134,83 +133,194 @@ func (r *RpcLogStreamer) Stop() {
+		return
+	}
 
-	logger.Info("Starting watcher", zap.Uint64("fromBlock", fromBlock))
+	innerBackfillCh, cancelBackfill, err := r.buildBackfillChannel(cfg, backfillStartBlock)
+	if err != nil {
+		logger.Error("failed to retrieve historical logs", zap.Error(err))
+		return
+	}
+
+	innerSubCh, err := r.buildSubscriptionChannel(cfg, subLogger)
+	if err != nil {
+		logger.Error("failed to subscribe to contract", zap.Error(err))
+		return
+	}
+
+	logger.Info("Starting watcher")
 
 	for {
 		select {
 		case <-r.ctx.Done():
-			logger.Debug("Stopping watcher")
+			logger.Debug("Context cancelled, stopping watcher")
 			return
-		case <-timer.C:
-			logger.Fatal(
-				"Max disconnect time exceeded. Node might drift too far away from expected state. Shutting down...",
-			)
 
-		case reorgBlock, open := <-watcher.reorgChannel:
+		case reorgBlock, open := <-cfg.reorgChannel:
 			if !open {
 				logger.Debug("Reorg channel closed")
 				return
 			}
-			fromBlock = reorgBlock
+			backfillStartBlock = reorgBlock
 			logger.Warn(
 				"Blockchain reorg detected, resuming from block",
-				zap.Uint64("fromBlock", fromBlock),
+				zap.Uint64("fromBlock", backfillStartBlock),
 			)
-			timer.Reset(watcher.MaxDisconnectTime)
-
-		default:
-			logs, nextBlock, err := r.GetNextPage(watcher, fromBlock)
-			if err != nil {
-				logger.Error(
-					"Error getting next page",
-					zap.Uint64("fromBlock", fromBlock),
-					zap.Error(err),
-				)
-				time.Sleep(ERROR_SLEEP_TIME)
+
+		case log, open := <-innerSubCh:
+			if !open {
+				subLogger.Fatal("subscription channel closed, closing watcher")
+				return
+			}
+
+			if !isSubEnabled {
+				// backfillEndBlock is a moving target.
+				// Subscription is always ahead of backfill, so we update backfillEndBlock when a new log is received.
+				if log.BlockNumber > backfillEndBlock {
+					backfillEndBlock = log.BlockNumber
+				}
+
+				// TODO: Next PR to introduce a log buffer.
+				// The buffer has to be of size lagFromHighestBlock, at minimum.
+
 				continue
 			}
-			// reset self-termination timer
-			timer.Reset(watcher.MaxDisconnectTime)
 
-			if nextBlock != nil {
-				fromBlock = *nextBlock
+			subLogger.Debug(
+				"Sending log to subscription channel",
+				zap.Uint64("fromBlock", log.BlockNumber),
+			)
+
+			cfg.eventChannel <- log
+
+		case log, open := <-innerBackfillCh:
+			if !open {
+				innerBackfillCh = nil
+				continue
 			}
 
-			if len(logs) == 0 {
-				time.Sleep(NO_LOGS_SLEEP_TIME)
+			// This is a guard, this case shouldn't happen.
+			// When the subscription is enabled the backfill is always cancelled.
+			if isSubEnabled {
 				continue
 			}
 
-			logger.Debug(
-				"Got logs",
-				zap.Int("numLogs", len(logs)),
-				zap.Uint64("fromBlock", fromBlock),
-				zap.Time("time", time.Now()),
-			)
-			for _, log := range logs {
-				watcher.backfillChannel <- log
+			if log.BlockNumber > highestBackfillBlock {
+				highestBackfillBlock = log.BlockNumber
+			}
```

Comment:
can the backfill be out of order? Why would log.Blocknumber ever be lower than the one before?